iml-services/iml-device/src/virtual_device/mod.rs

```diff
@@ -0,0 +1,292 @@
+            ps.len(),
+            f.to_string(),
+            note
+        );
+        parents.extend(ps);
+    }
+
+    for (f, d) in devices2 {
+        let dd = insert_virtual_devices(d, &parents);
+
+        results.push((f, dd));
+    }
+
+    results
+}
+
+fn collect_virtual_device_parents(
+    d: &Device,
+    level: usize,
+    parent: Option<&Device>,
+) -> Vec<Device> {
+    let mut results = vec![];
+
+    if is_virtual(d) {
+        tracing::debug!(
+            "Collecting parent {} of {}",
+            parent.map(|x| to_display(x)).unwrap_or("None".into()),
+            to_display(d)
+        );
+        vec![parent
+            .expect("Tried to push to parents the parent of the Root, which doesn't exist")
+            .clone()]
+    } else {
+        match d {
+            Device::Root(dd) => {
+                for c in dd.children.iter() {
+                    results.extend(collect_virtual_device_parents(c, level + 1, Some(d)));
+                }
+                results
+            }
+            Device::ScsiDevice(dd) => {
+                for c in dd.children.iter() {
+                    results.extend(collect_virtual_device_parents(c, level + 1, Some(d)));
+                }
+                results
+            }
+            Device::Partition(dd) => {
+                for c in dd.children.iter() {
+                    results.extend(collect_virtual_device_parents(c, level + 1, Some(d)));
+                }
+                results
+            }
+            Device::Mpath(dd) => {
+                for c in dd.children.iter() {
+                    results.extend(collect_virtual_device_parents(c, level + 1, Some(d)));
+                }
+                results
+            }
+            _ => vec![],
+        }
+    }
+}
+
+// This function accepts ownership of `Device` to be able to reconstruct
+// its `children`, which is an `OrdSet`, inside of `insert`.
+// `OrdSet` doesn't have `iter_mut` so iterating `children` and mutating them in-place isn't possible.
+fn insert_virtual_devices(mut d: Device, parents: &collections::HashSet<Device>) -> Device {
+    for p in parents {
+        d = insert(d, &p);
+    }
+    d
+}
+
+// This function in only for deduplication of `else` branch in `insert`
+fn new_children(d: Device, to_insert: &Device) -> OrdSet<Device> {
+    children_owned(d)
+        .into_iter()
+        .map(|c| insert(c, to_insert))
+        .collect()
+}
+
+fn insert(mut d: Device, to_insert: &Device) -> Device {
+    if compare_selected_fields(&d, to_insert) {
+        tracing::debug!(
+            "Inserting device {} children to {}",
+            to_display(to_insert),
+            to_display(&d)
+        );
+
+        children_mut(&mut d).map(|x| {
+            children(to_insert).map(|y| {
+                *x = y.clone();
+            })
+        });
+
+        d
+    } else {
+        // This is a shallow copy due to `children` being `im::OrdSet`, which is copy-on-write.
```

Comment:
Last time I tested this it didn't hold up (it could be different now). I added a number of structs with a custom clone impl that would print out when called. I then cloned the container and saw that the items in the container all were cloned as well.